File: Experiment-11-FLSM subnetting/flsm.py
import math
import ipaddress

network = input("Enter Network Address (e.g. 192.168.1.0/24): ")
subnets = int(input("Enter Number of Subnets: "))

net = ipaddress.ip_network(network, strict=False)

bits = math.ceil(math.log2(subnets))

new_prefix = net.prefixlen + bits

if new_prefix > 32:
    print("Invalid number of subnets.")
    exit()

print("\nSubnet Details:\n")

for i, subnet in enumerate(net.subnets(new_prefix=new_prefix), start=1):
    print(f"Subnet {i}")
    print("Network Address :", subnet.network_address)
    print("Broadcast Address :", subnet.broadcast_address)
    print("Subnet Mask :", subnet.netmask)
    print("Usable Hosts :", subnet.num_addresses - 2)
    print("-" * 40)

File: Experiment-11-FLSM subnetting/flsm.cpp
#include <iostream>
#include <cmath>
using namespace std;

int main()
{
    string network;
    int prefix;
    int subnets;

    cout << "Enter Network Address (Example: 192.168.1.0): ";
    cin >> network;

    cout << "Enter Prefix Length: ";
    cin >> prefix;

    cout << "Enter Number of Subnets: ";
    cin >> subnets;

    int bitsBorrowed = ceil(log2(subnets));

    int newPrefix = prefix + bitsBorrowed;

    int blockSize = pow(2, 32 - newPrefix);

    cout << "\nNew Prefix = /" << newPrefix << endl;
    cout << "Block Size = " << blockSize << endl;

    // Extract first three octets
    int a,b,c,d;
    sscanf(network.c_str(), "%d.%d.%d.%d",&a,&b,&c,&d);

    cout << "\nSubnet Details\n";

    int start = 0;

    for(int i=1;i<=subnets;i++)
    {
        cout << "\nSubnet " << i << endl;

        cout << "Network Address : "
             << a<<"."<<b<<"."<<c<<"."<<start<<endl;

        cout << "Broadcast Address : "
             << a<<"."<<b<<"."<<c<<"."<<(start+blockSize-1)<<endl;

        cout << "Usable Host Range : "
             << a<<"."<<b<<"."<<c<<"."<<(start+1)
             <<" - "
             <<a<<"."<<b<<"."<<c<<"."<<(start+blockSize-2)
             <<endl;

        start += blockSize;
    }

    return 0;
}